### ISSUE TITLE
DSpark speculative decoding support

### ISSUE THREAD

**Issue by notatestuser** [notatestuser/ds4-control#9]
Add opt-in support for **DSpark**, DeepSeek's auxiliary draft model for V4 Flash. It reads hidden states from the main model, proposes up to five future tokens, and Flash verifies them — committing only the accepted prefix. Shipped upstream as a separate ~5.6 GiB support GGUF (`ds4f-dspark` → `DeepSeek-V4-Flash-DSpark-support-0731.gguf`), passed to ds4-server as `--mtp <file> --dspark`.

## Scope

- Settings toggle, off by default (it needs a separate 5.6 GiB download and is experimental upstream).
- Independent background download track with progress in the Settings pane — it must **not** enter `ServerState.downloading`, so Start/Stop and the popup stay usable while a ~5.6 GiB accessory downloads.
- "Remove download" affordance to reclaim the space.
- Pass `--mtp <path> --dspark` on next start/restart.

## Engine constraints that shape the design

`ds4-server` only engages speculation when **all** of these hold (`ds4_server.c`):

```c
if (!s->batched_mode && temperature <= 0.0f &&
    ds4_engine_mtp_draft_tokens(s->engine) > 1 && ...)
```

1. **Not batched** — `--batched-session` (Concurrent sessions > 1) disables speculation outright.
2. **`temperature <= 0`** — greedy only. ds4's only speculative entry point is `ds4_session_eval_speculative_argmax`; the drafter argmaxes its own logits and verification is accept-to-first-mismatch, so there are no probabilities to do rejection sampling with.
3. **V4 Flash only** — PRO is unsupported, and `--dspark` without a readable `--mtp FILE` is a hard engine-open failure.

So the app must only pass the flags when they can actually work, and surface why when it can't.

## Open design point

Because ds4 skips speculation above temperature 0, and the built-in chat sends 0.7, wiring up the flags alone ships a toggle that does nothing for the only client we control.

DeepSeek's V4 model card recommends `temperature = 1.0` / `top_p = 1.0` (ds4 implements exactly that as its defaults), so forcing the chat to 0 trades recommended sampling — and, on a thinking model, carries repetition risk — for a speedup upstream itself calls experimental.

**Resolution: decouple.** DSpark ships as its own toggle, and greedy chat becomes a *separate*, off-by-default "Greedy chat replies" switch, so the trade is knowing rather than silent.

## Status

Implemented on `feat/dspark-speculative-decoding`.

**Comment by notatestuser** [comment]
## Benchmark: DSpark is a net loss here, and it changes greedy output

**Setup** — ds4 `ebacae9` (upstream `84cc882` + our THINK_MAX patch, i.e. PR #10's engine), V4 Flash `q4-imatrix`, M3 Ultra 512 GB, `--ctx 32768`, Metal, single session, default confidence (Metal 0.6). All requests `temperature: 0`, `max_tokens: 400`. Speed is the server's own per-request decode counter, median of warm reps.

Support model loaded clean: `stages=3 block=5 markov_rank=256 tensors=81 missing=0 invalid=0`.

### Speed — helps code slightly, hurts everything else

| category | baseline | DSpark | delta |
|---|---:|---:|---:|
| code (rbtree) | 41.42 | 42.38 | **+2.3%** |
| code (lru) | 41.34 | 42.17 | **+2.0%** |
| structured (tcp) | 41.22 | 37.25 | −9.7% |
| structured (gitflow) | 41.12 | 36.41 | −11.5% |
| creative (story) | 41.15 | 36.47 | −11.4% |
| creative (poem) | 40.95 | 36.99 | −9.7% |
| **overall** | **41.20** | **38.61** | **−6.3%** |

This matches upstream's warning ("predictable continuations, especially code, benefit most; low-yield prompts can be no faster or even slower") — but the code gain is ~2%, while the penalty elsewhere is ~10%.

### Why: the engine's own accounting says it doesn't pay

```
cycles=3794 proposed=4409 accepted_draft=3272 accept_rate=74.21% avg_accept=0.862
no_draft=2257 scheduler_skips=1813 miss_first=241
accepted_len_hist=  0:2498  1:413  2:328  3:222  4:128  5:205
propose=15635ms  verify=75036ms  saved=79474ms  net_saved=-12016ms
```

`accept_rate` looks healthy at 74%, but that is *of tokens actually proposed*. The number that matters is **`avg_accept=0.862`** — fewer than one draft token accepted per cycle, when each cycle costs a propose *and* a verify. 2498 of 3794 cycles accepted **zero**, and 2257 produced no draft at all.

**`net_saved = −12.0 s`** is the engine's own verdict: across the run, speculation cost 12 seconds more than it saved. The scheduler is already defending against this (1813 skips); without it the loss would be larger.

### Greedy identity is broken

At `temperature: 0`, DSpark should be lossless w.r.t. ordinary greedy decode. It isn't — **5 of 6 prompts produced different text**.

Controls, to rule out my harness:

| check | result |
|---|---|
| baseline repeated 3× | identical |
| baseline across a server restart | identical |
| **DSpark repeated 3×** | **identical** |
| **DSpark vs baseline, same prompt** | **differs** |

So both paths are individually deterministic — this is a systematic decode difference, not flakiness. Divergence is substantive, not whitespace:

```
baseline: ...from dataclasses import dataclass
dspark  : ...from collections import OrderedDict
```

Worth noting upstream carries a commit `Restore greedy identity after DSpark acceptance`, so this property is intended to hold. On this build it does not.

### Verdict

Keeping the toggle **off by default** is correct, and the separate `Greedy replies` opt-in was the right call — coupling them would have traded ~10% throughput plus changed output for a 2% gain on code.

Nothing to change in the branch. If DSpark is revisited, the greedy-identity failure is the blocker to chase first; the speed profile is secondary and may improve as upstream tunes the scheduler.

<sub>Reproduce: `DS4_DSPARK_STATS=1 ds4-server -m <q4> --ctx 32768 --metal --mtp <support>.gguf --dspark`, stats print on SIGTERM.</sub>

**Comment by pauleveritt** [comment]
As a note, I also saw negative numbers when first trying it. I know antirez did a number of commits in the last couple of days, wasn't sure if that changed the story.
